src/Simulation/Common/QubitManager.cs

```diff
@@ -81,20 +81,20 @@ protected class StackFrame
         {
             internal IApplyData? Data;
             internal IEnumerable<Qubit>? QubitsInArgument => Data?.Qubits;
-            private List<Qubit>? _locals; // Qubits allocated/borrowed in the current operation
+            private HashSet<Qubit>? _locals; // Qubits allocated/borrowed in the current operation
 
             public StackFrame(IApplyData data = null)
             {
                 Data = data;
             }
 
-            public List<Qubit> Locals
+            public HashSet<Qubit> Locals
             {
                 get
                 {
                     if (_locals == null)
                     {
-                        _locals = new List<Qubit>();
+                        _locals = new HashSet<Qubit>();
```

Comment:
Is Qubit class or record (i.e. by reference or by value)? Does default hash provider work well?

Reply:
Qubit is a class and overrides GetHashCode (https://github.com/microsoft/qsharp-runtime/blob/main/src/Simulation/Core/Qubit.cs)